applications/tari_base_node/src/parser.rs

```diff
@@ -1686,16 +1690,64 @@ impl Parser {
         }
 
-        // TODO: Implement Tx rate vs. as fast as possible, must be non-blocking
-        // TODO: Start at specified time, must be non-blocking
-        for i in 0..(tx_per_s * duration as f64) as usize {
-            // `send-tari` commands: [amount of tari to send] [destination public key or emoji id] [optional: msg]
-            let command_str =
-                (start_amount.0 + amount_inc.0 * i as u64).to_string() + " " + &dest_pubkey.to_string() + " " + &msg;
-            let args = command_str.split_whitespace();
-            // Execute
-            self.process_send_tari(args);
-        }
+        let mut dht = self.discovery_service.clone();
+        let executor = self.executor.clone();
+        let wallet_transaction_service = self.wallet_transaction_service.clone();
+        self.executor.spawn(async move {
+            // Ensure a valid connection is available by forcing a peer discovery. This is intended to be
+            // a blocking operation before the test starts.
+            match dht
+                .discover_peer(
+                    Box::from(dest_pubkey.clone()),
+                    NodeDestination::PublicKey(Box::from(dest_pubkey.clone())),
+                )
+                .await
+            {
+                Ok(_p) => {
+                    // Wait until specified test start time
+                    let millis_to_wait = (time_utc_start - Utc::now()).num_milliseconds();
+                    println!(
+                        "`make-it-rain` to peer '{}' scheduled to start at {}: msg \"{}\"",
+                        &key, time_utc_start, &msg
+                    );
+                    if millis_to_wait > 0 {
+                        tokio::time::delay_for(Duration::from_millis(millis_to_wait as u64)).await;
+                    }
+
+                    // Send all the transactions
+                    let start = Utc::now();
+                    for i in 0..number_of_txs {
+                        // `send-tari` commands: [amount of tari to send] [destination public key or emoji id]
+                        // [optional: msg]
+                        let command_str = (start_amount.0 + amount_inc.0 * i as u64).to_string() +
+                            " " +
+                            &dest_pubkey.to_string() +
+                            " " +
+                            &msg;
+                        let args = command_str.split_whitespace();
+                        // Execute
+                        let millis_actual_i = (Utc::now() - start).num_milliseconds() as u64;
+                        let millis_target_i = (i as f64 / (tx_per_s / 1000.0)) as u64;
+                        if millis_target_i - millis_actual_i > 0 {
+                            tokio::time::delay_for(Duration::from_millis(millis_target_i - millis_actual_i)).await;
+                        }
+                        Self::process_send_tari(args, executor.clone(), wallet_transaction_service.clone());
```

Comment:
Would it not be more ergonomic and more intuitive to rather move the send logic out from `process_send_tari(..)`
And then let make_it_rain and process_send_tari call that function. 

Currently here you have to create "fake" arguments and then reparse them again to actual arguments.

Reply:
Hi thank you for the comment. I do not agree for the following reasons:
- This is a specific variation for `process_send_tari`, and a custom command must be created and passed to the function, thus no duplication.
- W.r.t. the send logic, the `process_make_it_rain` function uses all the send logic of the `process_send_tari`, but not vice versa.